### ISSUE TITLE
which 40b weight should be promoted?

### ISSUE THREAD

**Issue by l1t1** [gcp/leela-zero#1798]
let e2be be the benchmark, many new weights only got 0.51 winrate,  99b6 got 0.528, but a304 is slight better than e2be, while b072 beat a304 by 0.53...

**Comment by l1t1** [comment]
but in fact b072 is quantized a304...

**Comment by wonderingabout** [comment]
i think we are fine choosing whatever 40b that has 52+% winrate against e2be, and if they both have about the same winrate, i'd rather go with the most recent (newest data)

about b072, it is true that it got 53% against a304 which got 50+% against e2be, however looking at the match games data, it seems a304 got a lucky boost at the start and never managed to increase its winrate against e2be later in the match, so i'm assuming that a304's winrate against e2be is possibly slightly overestimated

lets imagine that the real winrate of a304 vs e2be would be 49%, then the 53.3% that b07 value about the same than the 52-53% that other 40b got against e2be

but again, since they are about the same strength, i'd just go with the most recent one having a high enough winrate against e2be
what matters is not really the starting point, but what happens after it

**Comment by Marcin1960** [comment]
@wonderingabout "i think we are fine choosing whatever 40b that has 52+% winrate"

Here is a utility that randomly simulated winrates:
[NetChance.zip](https://github.com/gcp/leela-zero/files/2343391/NetChance.zip)

**Comment by wonderingabout** [comment]
@Marcin1960 
i saw you mention it in another post, i like this funny and original idea

@bubblesld 
as was said in #1791 and #1797 , elf v1 is not a reliable benchmark to estimate leela zero 40b strength, because @bjiyxo said, leela zero 40b and 20b has learned most of the "tricks" and knowledge of elf v1, but elf v1 has still a (much) stronger middle and endgame
so many elf v1 early resigned games (arround move 100) are games estimates it lost them due to its sharp winrate, but leela zero wouldnt be able to actually play them all until they are won, because elf v1 has a stronger middle game
so, as i said in another issue, to compare 40b's strength, it is better to use another 40b network, and we used mostly e2be so far, so keeping it the universal benchmark is reliable

again, elf v1 match game results are not reliable and cant be interpreted because of the bias and uncertainties i mentionned earlier

**Comment by bubblesld** [comment]
@wonderingabout 

As I mentioned in the previous reply, the winning rate from 47%-53% means little. The difference is usually due to luck than strength. It does not make sense to find one with the highest w.r. against e2b3, but only about 53%. If we can find one with 55+%, we would like to promote it for sure.

On the other hand, I do not agree that the results vs elf1 are unreliable. Yes, 40b learns some opening tricks to beat efl1 early, and it is weaker than elf1 in the later stage. But it is not just luck. It is some learned trick. Some weights are more familiar with the trick, while some are not. People may find weights with description 40b-167-xxxk and 40b-167a-xxxk. The reason we retrained the cycle up to v 167 is that we found that the produced weights become much weaker against elf1. Then we increased the percentage of elf type games from 30% to 40%. Then the produced weights can maintain a w.r. vs elf1 at around 50%.

What I want to state is: the training games are from 20b and elf1. It is difficult to improve the w.r. of match games among 40b weights. What we can evaluate are from games vs 20b and elf1 (see if 40b learns tricks to beat them)

**Comment by cjohnchen** [comment]
As I tested, c9d7 is yje best

**Comment by cjohnchen** [comment]
As I tested, c9d7 is the best

**Comment by wonderingabout** [comment]
@bubblesld 
naturally, training on almost the same data, we cant expect drastic changes in winrate against e2be
thats why i think going with the latest one is fine, after all what matters is not the starting point but what happens after it

**Comment by gcp** [comment]
Which has the highest estimated Elo (against current 20b)?

Does the estimated best change if we average out the result against Elf?

In the end small differences shouldn't matter too much here.

**Comment by herazul** [comment]
Considering that the training of the 40b on the 20b self-play games almost didn't improve it, would'nt it be better to take one of the first one released (the e2be that had 50% wr against elfv1) ? I don't know if it makes sense, but if the net is "less trained", maybe it has higher chance to work well without having the highest learning rate ? (but maybe it doesn't make any sense i don't know)

**Comment by bjiyxo** [comment]
IMO, c9d70c41 is better and we should promote it immediately.

158c107e is a experimental version(https://github.com/gcp/leela-zero/issues/1681#issuecomment-417853723), but we are not sure if it is effective. Furthermore, it looks like 40b can hardly improve by training 20b & elf1's self-play. So promoting c9d70c41 immediately will be the best choice.

**Comment by herazul** [comment]
@bjiyxo how do you know that c9d7 is better than e2be ? they have similar winrate against each other. It's an interesting question, how do you measure a net strengh ? against elf v0 or v1 ? against 20b leela net ? against itself (vs previous iteration) ?

**Comment by Mardak** [comment]
Shall we run a normal promotion match c9d70c41 vs LZ173 33986b7f and get the lastest 40b vs 20b win rate?

**Comment by gcp** [comment]
I'll schedule it tomorrow morning, so I can adjust the training machine immediately.

**Comment by bubblesld** [comment]
@herazul 
As you mentioned, it is difficult to tell which one is better by the matches against each other. But c9d7 has a much better w.r. against elf1. There is no test for e2be against 20b. But from c9d7 has a better w.r. against v172 comparing to elf1 vs v172, we will guess that it is improving in matches against 20b. From all these, c9d7 is probably is safer choice. 

But I agree that the difference between 40b's is little. It is easier to progress by having 40b self-games.

**Comment by alphaladder** [comment]
It is a good choice to progress 40b by self play other than training by 20b &elf games.

**Comment by Mardak** [comment]
@gcp, should we turn off ELFv1 game generation around the **same time** (😱 )? I believe we had 280k ELFv0 (some with resign turned off) and we're at 530k total ELF, so we're right at the 250k target.

**Comment by alphaladder** [comment]
@mardak  I guess gcp would like to stop elf self play at the time of promoting 40b. :)

**Comment by gcp** [comment]
Graph looks a bit funny now but will sort itself out as the next trained 40b's appear.

**Comment by herazul** [comment]
There is a problem due to the fact that c9d7 games were uploaded by someone befor promotion a week ago or something like this, i think roy need to swap the net's id between 173 and 174 ^^
also what's the deal with a15c ? it seems like a strong net, maybe it needs a promotion match against c9d7 to start from the best 40b net this morning, if this net is indeed better than c9d7 ! (or maybe it was lucky for the first 13 games, we will see)

**Comment by l1t1** [comment]
a15c was lucky, but is not now

**Comment by l1t1** [comment]
ce00 would be the best 20b if no new 20b will test

**Comment by bubblesld** [comment]
It is not a16c is lucky in the beginning. The fact is that 40b tends to win games early vs elf1. Therefore, the early returned results are in favor 40b. Check [390a vs elf1](http://zero.sjeng.org/match-games/5b82cca1580ff446786c2b1d), 390a won 14 in a row from the beginning, but only has 46.8% in the end.

In my personal test, a15c was 43:28 against elf1. That is why we put it for test. But probably it is lucky in my test.

**Comment by l1t1** [comment]
b75d looks better, which vs elfv1 0.51, vs e2be 0.52
while
c9d7 vs e2be 0.49
both b75d and c9d7 are trained by data up to lz 167

**Comment by jkiliani** [comment]
@gcp What is the plan regarding the training pipeline now? Use very low learning rates as with the 20b net? Since this net was trained on really low learning rates, shifting to higher ones will in all likelihood prevent the trained nets from passing gating.

The only method I can think of to get the benefits of higher learning rates but still have nets that may pass gating would be to try cyclical schedules for a while, i.e. train a limited number of steps at the learning rates we used for the 15b net, and then use lower learning rates and SWA for the rest of the training run.

**Comment by roy7** [comment]
I changed the # of training games to make the graph look more reasonable, and I replaced the _id field on the old promoted network with a new _id so that the timestamp would order the network properly in the Networks table.

**Comment by roy7** [comment]
@gcp didn't answer @Mardak's question but I went ahead and turned off ELF selfplay.

**Comment by Friday9i** [comment]
@jkiliani It may indeed be difficult to go out from the local minima, with the 40b having being trained with a small Learning Rate... For medium term progress, we probably have to increase the Learning Rate, but will we manage to promote with a significantly higher LR ...?
If we try a significantly higher LR, we have 2 possibilities:
- The easy one: the high LR will make trained nets go out of local minima (hence, less performance) but if one manages to learn new concepts and that more than compensate the loss of perf, it will promote: following nets will also be out of local minima and so all is well, progresses should continue ;-)
- The hard one...: the loss of perf due to high LR is not compensated by new 40b concepts, and new nets do not manage to promote :-(... If that's the case, what to do? As stated above, lowering LR is probably not a good idea for medium term progress. Proposition: we could keep the higher LR and promote the first "reasonably weaker" net **in order to go out of the local minima**, then come back to the traditionnal LZ 55% gating approach.

**Comment by alphaladder** [comment]
@roy7 the start date of #174 should be 9-04-18, not 9-24-18, kindly reminder.

**Comment by jkiliani** [comment]
> @roy7 the start date of #174 should be not 2018-09-04. 05:00, not 2018-09-24 05:00.

I agree, dating the net into the future is asking for more problems when the next net promotion happens, presumably before the 24th of this month.

**Comment by l1t1** [comment]
I saw the training data of 090c3d4b is 10.01M+32.0k, how many games did it include and what are they(lz 172? elfv1?)?

**Comment by l1t1** [comment]
090c3d4b lose in a row
![2018-09-05_090917](https://user-images.githubusercontent.com/23301704/45065315-0efc0280-b0eb-11e8-880d-4453f45d88b3.png)

**Comment by l1t1** [comment]
suggest to use gz max compress on 40b weights, the diff is about 7~10 MB
@bjiyxo c9d7 103927180
bjiyxo 33a3 103933212
max gz  c9d7 93871262
offical 09c3 100338629

**Comment by gcp** [comment]
The weights from the training server are always compressed with gz max.

**Comment by l1t1** [comment]
I know the reason from https://www.7-zip.org/ 
•For ZIP and GZIP formats, 7-Zip provides a compression ratio that is 2-10 % better than the ratio provided by PKZip and WinZip

**Comment by jillybob** [comment]
Given decompressing files is not an issue for users (as network file can be loaded within a zipped file) 7zip compression makes more sense than gz

"In terms of availability, 7zip is widely available across UNIX (Linux/BSD/MacOS) and Windows systems. Therefore a zip file is highly portable. Tools to compress/decompress xz and gzip files are also available on Windows systems, but are more commonly seen and used on UNIX systems.

xz and 7zip are known to have a better compression algorithm than gzip, but use more memory and time to compress/decompress. This topic is nicely discussed here.

I would recommend using gzip when less memory is available, and compression/decompression speed is a concern. 7zip and xz can be used when space is a concern and compression/decompression speed is not.

Some nice benchmarks on these algorithms can be found here. Note: LZMA is the compression algorithm used by 7zip."

https://unix.stackexchange.com/questions/322746/7zip-xz-gzip-tar-etc-what-should-i-use

**Comment by gcp** [comment]
Using LZMA (which is really what 7z uses) is not even under consideration.

He's talking about using the deflate code in 7z which is a bit better than the one in gzip. (There's an even better one in zopfli, but IIRC it's very slow the compress).

**Comment by gcp** [comment]
I see about 6% gain from using 7zip's zip. Can probably use it on the training machine too.

**Comment by gcp** [comment]
>I saw the training data of 090c3d4b is 10.01M+32.0k, how many games did it include

250k of the last LZ games and 250k ELFv1

**Comment by gcp** [comment]
I started the 256x40 with 0.0001 @ bs=96, i.e. a bump up from what was working for the 256x20. We'll see what it gives here.

I'll also need to try to gradually phase out the ELFv1 data, especially if we get some promotions...

**Comment by TFiFiE** [comment]
I get even better compression with bzip2 than 7-Zip on the latest 40 blocks, despite being much faster, at least when not bothering with any options.

**Comment by l1t1** [comment]
lzma is slower and low compress than bsc when dealing with pure text file

D:\>c:bsc\bsc e D:\tool\leela15\networks\c9d70c413e589d338743bfa83783e23f378dc0b9aa98940a2fbd1d852dab8781 d:\lz174.bsc
This is bsc, Block Sorting Compressor. Version 3.1.0. 8 July 2012.
Copyright (c) 2009-2012 Ilya Grebnov <Ilya.Grebnov@gmail.com>.

D:\tool\leela15\networks\c9d70c413e589d338743bfa83783e2 compressed 369651545 into 68840564 in 16.427 seconds.

**Comment by alphaladder** [comment]
I started the 256x40 with 0.0001 @ bs=96, i.e. a bump up from what was working for the 256x20. We'll see what it gives here.
—————-/////
@gcp  Accoding to our data in the big network training . In your case, maybe the bs=96 is too small. His parameter is below:
 learning rate  0.0001, 
bs>256, 
256k  or 512k.

Just for your reference.

**Comment by bjiyxo** [comment]
LR is 0.0001 and BS is 256 in my case.

**Comment by alphaladder** [comment]
@roy7 the elo score of new b0a should be 12498, not 0

**Comment by l1t1** [comment]
lz 175 is lucky, 14vs9 in last 2e games

**Comment by roy7** [comment]
There's race condition in the graph cache for web site performance that if a promotion is right at game 400 (I think is the issue), the cached page won't refresh further until the next match starts. As soon as the next match queues up it should fix itself.
